src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectReloadManager.cs

```diff
@@ -0,0 +1,508 @@
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private IProjectThreadingService _threadHandling;
+        private IUserNotificationServices _userNotificationServices;
+
+        private uint  _solutionEventsCookie = VSConstants.VSCOOKIE_NIL;
+        private ITaskDelayScheduler _reloadDelayScheduler;
+        private const int ReloadDelay = 1000;   // delay 1s before applying updated project contents.
+
+        // Tracks the set of reloadable projects. The value is the file system watcher cookie
+        private Dictionary<IReloadableProject, uint> _registeredProjects = new Dictionary<IReloadableProject, uint>();
+
+        // Tracks the list of projects that have changed and need to be processed
+        private List<IReloadableProject> _changedProjects = new List<IReloadableProject>();
+
+        [ImportingConstructor]
+        public ProjectReloadManager(IProjectThreadingService threadHandling, [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, IUserNotificationServices userNotificationServices)
+            : base(threadHandling.JoinableTaskContext)
+        {
+            _threadHandling = threadHandling;
+            _serviceProvider = serviceProvider;
+            _userNotificationServices = userNotificationServices;
+        }
+
+        /// <summary>
+        /// Initialization occurs when the first project registers
+        /// </summary>
+        private Task Initialize()
+        {
+            return InitializeCoreAsync(CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Called by reloadable projects to register themselves
+        /// </summary>
+        public async Task RegisterProjectAsync(IReloadableProject project)
+        {
+            await Initialize().ConfigureAwait(false);
+            await _threadHandling.SwitchToUIThread();
+
+            RegisterProject(project);
+        }
+
+        public void RegisterProject(IReloadableProject project)
+        {
+            uint filechangeCookie;
+            _registeredProjects.TryGetValue(project, out filechangeCookie);
+            System.Diagnostics.Debug.Assert(filechangeCookie == VSConstants.VSCOOKIE_NIL);
+            if (filechangeCookie == VSConstants.VSCOOKIE_NIL)
+            {
+                IVsFileChangeEx fileChangeService = _serviceProvider.GetService<IVsFileChangeEx, SVsFileChangeEx>();
+                if (fileChangeService != null)
+                {
+                    int hr = fileChangeService.AdviseFileChange(project.ProjectFile, (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Time | _VSFILECHANGEFLAGS.VSFILECHG_Size), this, out filechangeCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr) && filechangeCookie != VSConstants.VSCOOKIE_NIL);
+                    _registeredProjects.Add(project, filechangeCookie);
```

Comment:
Similar to above, do we want to do a success check here, so release builds don't potential fail later if something went catastrophically wrong? Seems like we shouldn't add the project to the registered list without it actually succeeding.